### PR TITLE
feat: Add configurable retry timing for RunTransactionAsync

### LINKED ISSUE #10513: Google.Cloud.Firestore: Backoff on FirestoreDb.RunTransactionAsync retry
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

 **Is your feature request related to a problem? Please describe.**
Regarding [RunTransactionAsync](https://github.com/googleapis/google-cloud-dotnet/blob/main/apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs#L389) on the FirestoreDb class in the Google.Cloud.Firestore package.

When we call it with `MaxAttempts > 1` option, then we would like to have some sort of backoff policy, in our case for data contention, so that it does not spam the document when it is already in contention.
(note there is a `backoff` variable that is unused in the source code, not really relevant to the solution proposed below though)

 **Describe the solution you'd like**
Either a backoff policy similar to what we see in the `_batchGetCallSettings` field or being able to pass an optional backoff policy to the method / client.
There's the issue of backwards compatibility, so maybe it does need to configurable by the user, defaulting to current behavior (no backoff).

 **Describe alternatives you've considered**
Currently we're implementing our own backoff policy, wrapping around the firestoredb, but since you already have a backoff concept, it seemed reasonable to add it here as well.

 **Additional context**
I admit that I'm a bit unsure if there's some internal backoff policy inside the lower internal functions that I haven't been able to divine, but from what I could see there's nothing obvious going on, so please correct me if there's anything I've misunderstood.

### PR DESCRIPTION
**Pull request by jskeet**
There are further changes to come in terms of transaction retry, but this is at least a start. Note that this changes the default backoff from "none at all" to "100ms initial, with a multiplier of 1.3". That's a more reasonable default, and it seems unlikely that customers would actually *depend* on there being no backoff.

Fixes #10513

### CODE REVIEW

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs:421]: I think there's a bug here that was there before.

If commit fails with aborted and there are attempts left, we never rollback that failed transaction right? Or is it OK not to rollback it because it already failed?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs:428]: I think this is clearer inside the catch.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/TransactionOptions.cs
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/TransactionOptions.cs:33]: Could we deprecate this one instead of ignoring the one in RetrySettings?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs:428]: I'm generally nervous of doing stuff in catch blocks. There are some oddities about how code executes there - not that I can remember details. Let's chat next week.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs:421]: I think it's okay, but we'll need to check. We definitely need to make changes here in general though - this PR is only attempting to affect the timing.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/TransactionOptions.cs
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/TransactionOptions.cs:33]: Feels a bit like overkill if most people won't need to touch the timing - creating a retry setting is relatively complex. Happy to chat about it though.

**Comment by jskeet** [comment]
Revamped to just use an underlying RetrySettings - PTAL.

**Review by amanda-tarafa — APPROVED** [review]
LGTM, thanks!

**Comment by jskeet** [comment]
I've added tests, and this is now ready to merge if you're happy with it. (Again, it's not the final word on transaction retries - but it's at least an improvement in terms of timing.)

**Pull request merged**
